src/utils.c

```diff
@@ -2031,6 +2031,209 @@ gchar **utils_strv_join(gchar **first, gchar **second)
+ *
+ * @param strv The list of strings to process.
+ * @param num The number of strings contained in @a strv. Can be 0 if it's terminated by @c NULL.
+ *
+ * @return The common prefix that is part of all strings.
+ */
+static gchar *utils_strv_find_lcs(gchar **strv, gsize num)
+{
+	gchar *first, *_sub, *sub;
+	gsize n_chars;
+	gsize len;
+	gsize max = 0;
+	char *lcs;
+	gsize found;
+
+	if (strv == NULL)
+		return NULL;
+
+	first = strv[0];
+	len = strlen(first);
+
+	if (num == 0)
+		num = g_strv_length(strv);
+
+	/* sub is the working area where substrings from first are copied to */
+	sub = g_malloc(len+1);
+	lcs = g_strdup("");
+	foreach_str(_sub, first)
+	{
+		gsize chars_left = len - (_sub - first);
+		/* No point in continuing if the remainder is too short */
+		if (max > chars_left)
+			break;
+		for (n_chars = 1; n_chars <= chars_left; n_chars++)
+		{
+			g_strlcpy(sub, _sub, n_chars+1);
+			found = 1;
+			for (gsize i = 1; i < num; i++)
+			{
+				if (strstr(strv[i], sub) == NULL)
+					break;
+				found++;
+			}
+			if (found == num && n_chars > max)
+			{
+				max = n_chars;
+				SETPTR(lcs, g_strdup(sub));
+			}
+		}
+	}
+	g_free(sub);
+
+	return lcs;
+}
+
+
+/** Transform file names in a list to be shorter.
+ *
+ * This function takes a list of file names (probably with absolute paths), and
+ * transforms the paths such that they are short but still unique. This is intended
+ * for dialogs which present the file list to the user, where the base name may result
+ * in duplicates (showing the full path might be inappropriate).
+ *
+ * The algorthm strips the common prefix (e-g. the user's home directory) and
+ * replaces the longest common substring with an ellipsis ("...").
+ *
+ * @param file_names @array{length=num} The list of strings to process.
+ * @param num The number of strings contained in @a file_names. Can be 0 if it's terminated by @c NULL.
+ * @return @transfer{full} A newly-allocated array of transformed paths strings, terminated by
+            @c NULL. Use @c g_strfreev() to free it.
+ *
+ * @since 1.31 (API 232)
+ */
+GEANY_API_SYMBOL
+gchar **utils_strv_shorten_file_list(gchar **file_names, gsize num)
+{
+	gsize i;
+	gchar *prefix, *substring, *lcs;
+	gchar *start, *end;
+	gchar **names;
+	gsize prefix_len, lcs_len;
+
+	/* The return value shall have exactly the same size as the input. If the input is a
+	 * GStrv (last element is NULL), the output will follow suit. */
+	if (!num)
+		num = g_strv_length(file_names);
+	/* Always include a terminating NULL, enables easy freeing with g_strfreev() */
+	names = g_new0(gchar *, num + 1);
+
+	prefix = utils_strv_find_common_prefix(file_names, num);
+	/* First: determine the common prefix, that will be stripped.
+	 * Don't strip single-letter prefixes, such as '/' */
+	prefix_len = 0;
+	if (NZV(prefix) && prefix[1])
+	{
+		/* Only strip directory components, include trailing '/' */
+		start = strrchr(prefix, G_DIR_SEPARATOR);
```

Comment:
Meh, what's the point G_DIR_SEPARATOR if you need to handle both. Also, what about legitimate paths that contain the other slash (e.g. a file can contain the \ character on linux, not sure if the same is true for / on Windows).

Are you sure we don't have problems with this elsewhere inside Geany? I.e. is this something we support or not?

Reply:
> e.g. a file can contain the \ character on linux

Yes, handling of `\` has to be special-cased for Windows

> not sure if the same is true for / on Windows

I don't think so, IIUC both `\` and `/` are separators.  One simple non-optimized way to match it can be `"/" G_DIR_SEPARATOR_S` (e.g. preprocessor concatenation of either `"//"` or `"/\"`) together with things like `strpbrk()`.

> Are you sure we don't have problems with this elsewhere inside Geany? I.e. is this something we support or not?

I'm not entirely sure, but I think so.  However, we're probably normalizing the path at some point, which means we might not encounter the issue with paths we handle anyway.
So this might not be so important in the end.

Reply:
>  So this might not be so important in the end.

Ok, so I don't plan changes for now in this regard. Speak up if you chnage your mind.